packages/webdriverio/src/commands/browser/switchFrame.ts

```diff
@@ -274,36 +277,47 @@ export async function switchFrame (
      * the function for each of them.
      */
     if (typeof context === 'function') {
-        const allContexts = await sessionContext.getFlatContextTree()
-        const allContextIds = Object.keys(allContexts)
-        for (const contextId of allContextIds) {
-            const functionDeclaration = new Function(`
+        let foundContextId: string | undefined
+
+        await this.waitUntil(async () => {
+            const allContexts = await sessionContext.getFlatContextTree()
+            const allContextIds = Object.keys(allContexts)
+
+            for (const contextId of allContextIds) {
+                const functionDeclaration = new Function(`
                 return (${SCRIPT_PREFIX}${context.toString()}${SCRIPT_SUFFIX}).apply(this, arguments);
             `).toString()
-            const params: remote.ScriptCallFunctionParameters = {
-                functionDeclaration,
-                awaitPromise: false,
-                arguments: [],
-                target: { context: contextId }
-            }
+                const params: remote.ScriptCallFunctionParameters = {
+                    functionDeclaration,
+                    awaitPromise: false,
+                    arguments: [],
+                    target: { context: contextId }
+                }
 
-            const result = await this.scriptCallFunction(params).catch((err) => (
-                log.warn(`switchFrame context callback threw error: ${err.message}`)))
+                const result = await this.scriptCallFunction(params).catch((err) => {
+                    log.warn(`switchFrame context callback threw error: ${err.message}`)
+                    return undefined
+                })
 
-            if (!result || result.type !== 'success' || result.result.type !== 'boolean' || !result.result.value) {
-                continue
+                if (result && result.type === 'success' && result.result.type === 'boolean' && result.result.value) {
+                    foundContextId = contextId
+                    return true
+                }
             }
 
-            /**
-             * reset the context to the top level frame first so we can start the search from the root context
-             */
-            await browser.switchFrame(null)
-
-            await this.switchFrame(contextId)
-            return contextId
-        }
+            return false
+        }, {
+            timeout: this.options.waitforTimeout,
+            interval: this.options.waitforInterval,
+            timeoutMsg: 'Could not find the desired frame within the timeout'
+        })
 
-        throw new Error('Could not find the desired frame')
+        /**
+             * reset the context to the top level frame first so we can start the search from the root context
+         */
+        await this.switchFrame(null)
+        await this.switchFrame(foundContextId!)
```

Comment:
So we don't need this
```suggestion
        await this.switchFrame(foundContextId)
```